### ISSUE TITLE
Implementation of exponential extrapolator for the differentiable mitigate_with_zne functionality

### ISSUE THREAD

**Issue by cnktysz** [PennyLaneAI/pennylane#3123]
### Feature details

The newly added differentiable zero noise extrapolation `(mitigate_with_zne)` function only supports a polynomial extrapolation. mitiq equivalent of this function allows many other functions (e.g. exponential, linear, poly-exponential, etc. ([see](https://mitiq.readthedocs.io/en/v.0.1a2/guide/guide_05-factories.html))). I believe it would be a great addition to add these options. 

### How important would you say this feature is?

2: Somewhat important. Needed this quarter.

**Comment by cnktysz** [comment]
The implementation doesn't seem very complicated to me. I would like to start implementing with exponential fitter, but it would be great to hear the Pennylane core team's ideas.

**Comment by antalszava** [comment]
Hi @cnktysz, thanks for the feature request! :slightly_smiling_face:

I don't think there would be any blockers there, submissions would be welcome to add further extrapolators that support differentiation. :tada:

Let us know if you have have any questions with regards to contributing. In October, PennyLane is participating in [Hacktoberfest](https://hacktoberfest.com/) - if this is interesting, we can also tag this issue for the hackathon! :+1:

**Comment by Qottmann** [comment]
Hi @cnktysz good idea! As long as you can trace it back to a polynomial fit it should be relatively straight-forward.
I.e. if you mean by exponential fit $y=c_0\exp(c_1 x)$ then you can write a dummy function
```python
def exp_extrapolate(x, y):
    logy = qml.math.log(y)
    coef = _polyfit(x, logy, 1)
    return qml.math.exp(coef[1])
```
since $\log(y) = \log(c_1) + c_0 x$ and the extrapolated result is $y(0)=c_1$.

For extrapolation that cannot be reduced to polynomial fitting it gets more complicated. To preserve differentiability you have to stick to `pennylane.math` functions. Currently, the backbone in the ZNE is the custom `_polyfit` function (see [pennylane/transforms/mitigate.py#L247](https://github.com/PennyLaneAI/pennylane/blob/master/pennylane/transforms/mitigate.py#L247)), and you would have to write a more elaborate non-linear fitting function. 
This is totally possible, e.g. using backprop and doing a least-squares fit, but just a little more involved.

Alternatively, you can use any numpy/scipy fitting function if you can live without automatic differentiability!

**Comment by cnktysz** [comment]
Yes, improving performance of ZNE is my goal. I couldn't get good results with scaling factor up-to 3 and was thinking of using other extrapolators. A collaborator of mine said they got better performance with exponential fitter, that's why I was inclined to test it out and realized it was not implemented.

I think such intermediate scaling factors would work better, but as you said it is not straightforward to decide how to do this. I will definitely have a look at this direction as well. Thank you.

**Comment by cnktysz** [comment]
Hi @Qottmann. I implemented the extrapolator based on your suggestion. It works well on its own. Then, I tested it with the jax interface and unfortunately it couldn't trace the gradients. I realized the issue starts with the `qml.math.log` function. Seems like batched inputs do not go well with this function. I tried to find the source code for this function but can't seem to locate it in the pennylane repository. Can you point me to it if you know where it is defined. Thanks.

**Comment by josh146** [comment]
Hey @cnktysz! We are actually performing dynamic dispatch within the `qml.math` module :) the way it works is:

- If a math function is defined in `qml.math`, it defaults to this definition.
- If a math function is _not_ defined in `qml.math`, then we use [autoray](https://github.com/jcmgray/autoray) to attempt to automatically dispatch to the correct ML framework.

So in this case, `qml.math.log` would simply call `autoray.numpy.log` under the hood, and then autoray will determine which ML framework to call.
